`azurerm_route_server` - Add support for `hub_routing_preference` property

bubbletroubles opened the pull request:
<!--  All Submissions -->


## Community Note
<!-- Please leave the community note as is. -->
* Please vote on this PR by adding a :thumbsup: [reaction](https://blog.github.com/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/) to the original PR to help the community and maintainers prioritize for review
* Please do not leave comments along the lines of "+1", "me too" or "any updates", they generate extra noise for PR followers and do not help prioritize for review


## Description

<!-- Please include a description below with the reason for the PR, what it is doing, what it is trying to accomplish, and anything relevant for a reviewer to know. 

If this is a breaking change for users please detail how it cannot be avoided and why it should be made in a minor version of the provider -->

- Add support for the `hub_routing_preference` property in `azurerm_route_server`.
- Updated docs to sort properties in alphabetical order, and added new docs for the `hub_routing_preference` property.

Here are the references for this feature 
- https://learn.microsoft.com/en-us/azure/virtual-wan/about-virtual-hub-routing-preference
- https://learn.microsoft.com/en-us/rest/api/virtualwan/virtual-hubs/create-or-update?view=rest-virtualwan-2024-05-01&tabs=HTTP#hubroutingpreference

## PR Checklist

- [X] I have followed the guidelines in our [Contributing Documentation](../blob/main/contributing/README.md).
- [X] I have checked to ensure there aren't other open [Pull Requests](../pulls) for the same update/change.
- [X] I have checked if my changes close any open issues. If so please include appropriate [closing keywords](https://docs.github.com/en/issues/tracking-your-work-with-issues/linking-a-pull-request-to-an-issue#linking-a-pull-request-to-an-issue-using-a-keyword) below.
- [X] I have updated/added Documentation as required written in a helpful and kind way to assist users that may be unfamiliar with the resource / data source.
- [X] I have used a meaningful PR title to help maintainers and other users understand this change and help prevent duplicate work. 
For example: “`resource_name_here` - description of change e.g. adding property `new_property_name_here`”


<!-- You can erase any parts of this template below this point that are not applicable to your Pull Request. -->


## Changes to existing Resource / Data Source

- [X] I have added an explanation of what my changes do and why I'd like you to include them (This may be covered by linking to an issue above, but may benefit from additional explanation).
- [X] I have written new tests for my resource or datasource changes & updated any relevent documentation.
- [X] I have successfully run tests with my changes locally. If not, please provide details on testing challenges that prevented you running the tests.
- [ ] (For changes that include a **state migration only**). I have manually tested the migration path between relevant versions of the provider.


## Testing 

- [X] My submission includes Test coverage as described in the [Contribution Guide](../blob/main/contributing/topics/guide-new-resource.md) and the tests pass. (if this is not possible for any reason, please include details of why you did or could not add test coverage)

All tests are passing - see [here](https://dev.azure.com/bubbletroubles/terraform/_build/results?buildId=152&view=logs&j=275f1d19-1bd8-5591-b06b-07d489ea915a&t=33e5d3ec-87e7-5f80-0281-074c6962cb44).

```
==> Checking that code complies with gofmt requirements...
==> Checking that Custom Timeouts are used...
==> Checking that acceptance test packages are used...
TF_ACC=1 go test -v ./internal/services/network -run=TestAccRouteServer -parallel 6 -timeout 600m -ldflags="-X=github.com/hashicorp/terraform-provider-azurerm/version.ProviderVersion=acc"
=== RUN   TestAccRouteServerBgpConnection_basic
=== PAUSE TestAccRouteServerBgpConnection_basic
=== RUN   TestAccRouteServerBgpConnection_requiresImport
=== PAUSE TestAccRouteServerBgpConnection_requiresImport
=== RUN   TestAccRouteServer_basic
=== PAUSE TestAccRouteServer_basic
=== RUN   TestAccRouteServer_requiresImport
=== PAUSE TestAccRouteServer_requiresImport
=== RUN   TestAccRouteServer_complete
=== PAUSE TestAccRouteServer_complete
=== RUN   TestAccRouteServer_update
=== PAUSE TestAccRouteServer_update
=== CONT  TestAccRouteServerBgpConnection_basic
=== CONT  TestAccRouteServer_requiresImport
=== CONT  TestAccRouteServer_update
=== CONT  TestAccRouteServer_complete
=== CONT  TestAccRouteServer_basic
=== CONT  TestAccRouteServerBgpConnection_requiresImport
--- PASS: TestAccRouteServer_requiresImport (1211.88s)
--- PASS: TestAccRouteServer_basic (1221.75s)
--- PASS: TestAccRouteServerBgpConnection_requiresImport (1385.48s)
--- PASS: TestAccRouteServer_complete (1405.40s)
--- PASS: TestAccRouteServerBgpConnection_basic (1482.51s)
--- PASS: TestAccRouteServer_update (1518.60s)
PASS
ok  	github.com/hashicorp/terraform-provider-azurerm/internal/services/network	1518.659s

Finishing: make acctests
```

<!-- Please include testing logs or evidence here or an explanation on why no testing evidence can be provided. 

For state migrations please test the changes locally and provide details here, such as the versions involved in testing the migration path. For further details on testing state migration changes please see our guide on [state migrations](https://github.com/hashicorp/terraform-provider-azurerm/blob/main/contributing/topics/guide-state-migrations.md#testing) in the contributor documentation. -->


## Change Log

Below please provide what should go into the changelog (if anything) conforming to the [Changelog Format documented here](../blob/main/contributing/topics/maintainer-changelog.md).

<!-- Replace the changelog example below with your entry. One resource per line. -->

* `azurerm_route_server` - support for the `hub_routing_preference` property [GH-28363]


<!-- What type of PR is this? -->
This is a (please select all that apply):

- [ ] Bug Fix
- [ ] New Feature (ie adding a service, resource, or data source)
- [X] Enhancement
- [ ] Breaking Change



> [!NOTE] 
> If this PR changes meaningfully during the course of review please update the title and description as required.

stephybun commented on internal/services/network/route_server_resource.go:
```diff
@@ -83,6 +83,17 @@ func resourceRouteServer() *pluginsdk.Resource {
+			"hub_routing_preference": {
+				Type:     pluginsdk.TypeString,
+				Optional: true,
+				Computed: true,
```
Can we set a default value here instead of `Computed`? Or does the default value depend on the route server configuration?

bubbletroubles replied:
When I didn't have this set to Computed, the tests were failing. The issue was

- It would set the property (e.g. `hub_routing_preference = "ASPath"`)
- It would run `terraform apply`
- The property would be set correctly - so far so good
- It would then remove the `hub_routing_preference` property from the test config
- It then run `terraform plan`
- It says it will perform the following
```
          # azurerm_route_server.test will be updated in-place
          ~ resource "azurerm_route_server" "test" {
              - hub_routing_preference           = "ExpressRoute" -> null
                id                               = "<resource-id>"
                name                             = "acctestrs-241221104344729540"
                # (9 unchanged attributes hidden)
            }
```
- But I get the error ` testcase.go:173: Step 1/6 error: After applying this test step, the non-refresh plan was not empty.`

The way the API works is that it always needs to be set as either `ExpressRoute`, `ASPath` or `VPNGateway`. The default is ExpressRoute [as per the MS Docs](https://learn.microsoft.com/en-us/azure/virtual-wan/about-virtual-hub-routing-preference#:~:text=ExpressRoute%20(This%20is%20the%20default%20setting)), but we can't assume its ExpressRoute as customers may have changed it manually before support for the property exists. 

In this scenario, is `Computed` appropriate? Or is there a better way to handle it?

stephybun replied:
I understand the desire to make the introduction of this property accommodate users that have had to use workarounds to set this property outside of Terraform, but if there is a publicly documented as well as a behavioural default for a property like we do in this case then we should be setting that as the default over setting it to computed.

Computed is typically reserved for behaviours where the default value changes dependent on the resource configuration/tenant/location etc.

bubbletroubles replied:
Thanks stephybun - this has been updated in the latest commits. Ready for you to review :)

stephybun commented on internal/services/network/route_server_resource.go:
```diff
@@ -132,6 +143,7 @@ func resourceRouteServerCreate(d *pluginsdk.ResourceData, meta interface{}) erro
 		Properties: &virtualwans.VirtualHubProperties{
 			Sku:                        pointer.To(d.Get("sku").(string)),
 			AllowBranchToBranchTraffic: pointer.To(d.Get("branch_to_branch_traffic_enabled").(bool)),
+			HubRoutingPreference:       (*virtualwans.HubRoutingPreference)(pointer.To(d.Get("hub_routing_preference").(string))),
```
Can we change this to 
```suggestion
			HubRoutingPreference:       pointer.To(virtualwans.HubRoutingPreference(d.Get("hub_routing_preference").(string))),
```

stephybun commented on internal/services/network/route_server_resource.go:
```diff
@@ -209,6 +221,14 @@ func resourceRouteServerUpdate(d *pluginsdk.ResourceData, meta interface{}) erro
 
+	if d.HasChange("hub_routing_preference") {
+		if v, ok := d.GetOk("hub_routing_preference"); ok {
+			payload.Properties.HubRoutingPreference = (*virtualwans.HubRoutingPreference)(pointer.To(v.(string)))
```
```suggestion
			payload.Properties.HubRoutingPreference = pointer.To(virtualwans.HubRoutingPreference(v.(string))),
```

stephybun commented on internal/services/network/route_server_resource.go:
```diff
@@ -209,6 +221,14 @@ func resourceRouteServerUpdate(d *pluginsdk.ResourceData, meta interface{}) erro
+		if v, ok := d.GetOk("hub_routing_preference"); ok {
+			payload.Properties.HubRoutingPreference = (*virtualwans.HubRoutingPreference)(pointer.To(v.(string)))
+		} else {
+			payload.Properties.HubRoutingPreference = nil
```
Does this reset the preference back to a default?

bubbletroubles replied:
It will leave the value as it was last set and the `hub_routing_preference` will no longer be managed via terraform. Is this the best way to achieve this, or is there a more preferred way?

bubbletroubles replied:
Hi stephybun - I've make this more in line with other properties in the latest commit. Ready for your re-review.

stephybun commented on internal/services/network/route_server_resource.go:
```diff
@@ -272,6 +292,9 @@ func resourceRouteServerRead(d *pluginsdk.ResourceData, meta interface{}) error
 			if props.AllowBranchToBranchTraffic != nil {
 				d.Set("branch_to_branch_traffic_enabled", props.AllowBranchToBranchTraffic)
 			}
+			if props.HubRoutingPreference != nil {
+				d.Set("hub_routing_preference", string(*props.HubRoutingPreference))
+			}
```
This can be simplified to 
```suggestion
			d.Set("hub_routing_preference", pointer.From(props.HubRoutingPreference))
```

stephybun commented on internal/services/network/route_server_resource_test.go:
```diff
@@ -71,13 +72,15 @@ func TestAccRouteServer_update(t *testing.T) {
 			Check: acceptance.ComposeTestCheckFunc(
-				check.That(data.ResourceName).ExistsInAzure(r)),
+				check.That(data.ResourceName).ExistsInAzure(r),
+			),
```
Can we add another step updating the config back to `basic` to test the behaviour on line 228
https://github.com/hashicorp/terraform-provider-azurerm/blob/83444612400941ff4f4aa167eff17e29799e5f46/internal/services/network/route_server_resource.go#L228

bubbletroubles replied:
Updated the test to include config back to basic.

stephybun commented on website/docs/r/route_server.html.markdown:
```diff
@@ -65,23 +66,25 @@ The following arguments are supported:
 
--> **NOTE:** Azure Route Server requires a dedicated subnet named RouteServerSubnet. The subnet size has to be at least /27 or short prefix (such as /26 or /25) and cannot be attached to any security group, otherwise, you'll receive an error message when deploying the Route Server
+* `public_ip_address_id` - (Required) The ID of the Public IP Address. This option is required since September 1st 2021. Changing this forces a new resource to be created.
 
 * `sku` - (Required) The SKU of the Route Server. The only possible value is `Standard`. Changing this forces a new resource to be created.
 
-* `public_ip_address_id` - (Required) The ID of the Public IP Address. This option is required since September 1st 2021. Changing this forces a new resource to be created.
+* `subnet_id` - (Required) The ID of the Subnet that the Route Server will reside. Changing this forces a new resource to be created.
```
We appreciate the alphabetical ordering here, but the pattern we try to follow in the docs is to list all `Required` properties first in alphabetical order (except for name, resource group, location), then all `Optional` properties in alphabetical order.

bubbletroubles replied:
Thanks - I was following the guidance [Guide: Extending a Resource - Docs](https://github.com/hashicorp/terraform-provider-azurerm/blob/main/contributing/topics/guide-new-fields-to-resource.md#docs) which says to use the same order as the Schema, which is alphabetical [Guide: Extending a Resource - Schema](https://github.com/hashicorp/terraform-provider-azurerm/blob/main/contributing/topics/guide-new-fields-to-resource.md#schema).

I'll update this PR as per your suggestion.

bubbletroubles commented:
@stephybun thanks for the review. I've updated the code as per your suggestions and added the extra test. There are still a couple of open comments - let me know how you want them addressed.

Updated tests are passing - results are [here](https://dev.azure.com/bubbletroubles/terraform/_build/results?buildId=155&view=logs&j=275f1d19-1bd8-5591-b06b-07d489ea915a&t=33e5d3ec-87e7-5f80-0281-074c6962cb44)

bubbletroubles commented:
Hi @stephybun I've updated the PR to remove the computed property and fix the update logic. It's ready for you to re-review.

Test are passing - https://dev.azure.com/bubbletroubles/terraform/_build/results?buildId=177&view=logs&j=275f1d19-1bd8-5591-b06b-07d489ea915a&t=33e5d3ec-87e7-5f80-0281-074c6962cb44

stephybun commented on website/docs/r/route_server.html.markdown:
```diff
@@ -65,23 +66,25 @@ The following arguments are supported:
+* `branch_to_branch_traffic_enabled` - (Optional) Whether to enable route exchange between Azure Route Server and the gateway(s).
 
-* `branch_to_branch_traffic_enabled` - (Optional) Whether to enable route exchange between Azure Route Server and the gateway(s)
+* `hub_routing_preference` - (Optional) The hub routing preference. Valid values are `ASPath`, `ExpressRoute` or `VpnGateway`.
```
```suggestion
* `hub_routing_preference` - (Optional) The hub routing preference. Valid values are `ASPath`, `ExpressRoute` or `VpnGateway`. Defaults to `ExpressRoute`.
```

stephybun reviewed (commented):
A final comment on the docs, then this should be good to go!

bubbletroubles commented:
Thanks @stephybun I've applied your suggestion for the docs. Should be good to go.

stephybun reviewed (approved):
Thanks for this @bubbletroubles LGTM 🚀
